File: sdk/python/symbolchain/RuleBasedTransactionFactory.py
import functools
import inspect
import operator
from enum import Enum, Flag

from .BaseValue import BaseValue
from .ByteArray import ByteArray
from .TransactionDescriptorProcessor import TransactionDescriptorProcessor


def _name_to_enum_value(mapping, enum_type, enum_value_name):
	if enum_value_name not in mapping:
		raise ValueError(f'unknown value {enum_value_name} for type {enum_type}')

	return mapping[enum_value_name]


def _build_type_hints_map(struct_value):
	type_hints = {}
	for key, hint in struct_value.TYPE_HINTS.items():
		rule_name = None
		if hint.startswith('array['):
			rule_name = hint
		elif hint.startswith('enum:'):
			rule_name = hint[len('enum:'):]
		elif hint.startswith('pod:'):
			rule_name = hint[len('pod:'):]
		elif hint.startswith('struct:'):
			rule_name = hint

		if rule_name:
			type_hints[key] = rule_name

	return type_hints


def _type_converter_factory(module, custom_type_converter, value):
	if custom_type_converter and custom_type_converter(value):
		return custom_type_converter(value)

	if isinstance(value, ByteArray):
		return getattr(module, type(value).__name__)(value.bytes)

	return value


class RuleBasedTransactionFactory:
	"""Rule based transaction factory."""

	def __init__(self, module, type_converter=None, type_rule_overrides=None):
		"""Creates a rule based transaction factory for use with catbuffer generated code."""
		self.module = module
		self.type_converter = lambda value: _type_converter_factory(self.module, type_converter, value)
		self.type_rule_overrides = type_rule_overrides or {}
		self.rules = {}

	def _get_module_class(self, name):
		return getattr(self.module, name)

	def add_pod_parser(self, name, pod_class):
		"""Creates wrapper for SDK POD types."""

		if pod_class in self.type_rule_overrides:
			# if an override has been set for the type, use it
			self.rules[name] = self.type_rule_overrides[pod_class]
			return

		self.rules[name] = lambda value: pod_class(value) if not isinstance(value, pod_class) else value

	def add_flags_parser(self, name):
		"""Creates flag type parser."""
		flags_class = self._get_module_class(name)
		string_to_enum = dict(map(lambda key: (key.name.lower(), key), flags_class))

		def parser(flags):
			if isinstance(flags, str):
				enum_array = list(map(lambda flag_name: _name_to_enum_value(string_to_enum, name, flag_name), flags.split(' ')))
				return functools.reduce(operator.or_, enum_array)

			if isinstance(flags, int):
				return flags_class(flags)

			return flags

		self.rules[name] = parser

	def add_enum_parser(self, name):
		"""Creates enum type parser."""
		enum_class = self._get_module_class(name)
		string_to_enum = dict(map(lambda key: (key.name.lower(), key), enum_class))

		def parser(enum_value):
			if isinstance(enum_value, str):
				return _name_to_enum_value(string_to_enum, name, enum_value)

			if isinstance(enum_value, int):
				return enum_class(enum_value)

			return enum_value

		self.rules[name] = parser

	def add_struct_parser(self, name):
		"""Creates struct parser (to allow nested parsing)."""
		struct_class = self._get_module_class(name)

		def parser(struct_descriptor):
			struct_processor = self._create_processor(struct_descriptor)
			struct_value = struct_class()

			all_type_hints = _build_type_hints_map(struct_value)
			struct_processor.set_type_hints(all_type_hints)

			struct_processor.copy_to(struct_value)
			return struct_value

		self.rules[f'struct:{name}'] = parser

	def add_array_parser(self, name):
		"""Creates array type parser, based on some existing element type parser."""
		element_rule = self.rules[name]
		element_name = name[len('struct:'):] if name.startswith('struct:') else name

		def parser(values):
			return list(map(element_rule, values))

		self.rules[f'array[{element_name}]'] = parser

	def autodetect(self):
		"""Autodetects rules using reflection."""
		for class_name in dir(self.module):
			cls = getattr(self.module, class_name)
			if not inspect.isclass(cls):
				continue

			if issubclass(cls, BaseValue) and cls != BaseValue:
				self.add_pod_parser(class_name, cls)

			if issubclass(cls, Enum) and cls != Enum and cls != Flag:
				if issubclass(cls, Flag):
					self.add_flags_parser(class_name)
				else:
					self.add_enum_parser(class_name)

	def create_from_factory(self, factory, descriptor):
		"""Creates an entity from a descriptor using a factory."""
		processor = self._create_processor(descriptor)
		entity_type = processor.lookup_value('type')
		entity = factory(entity_type)

		all_type_hints = _build_type_hints_map(entity)
		processor.set_type_hints(all_type_hints)
		processor.copy_to(entity, ['type'])

		self._auto_encode_strings(entity)
		return entity

	def _create_processor(self, descriptor):
		return TransactionDescriptorProcessor(descriptor, self.rules, self.type_converter)

	@staticmethod
	def _auto_encode_strings(entity):
		for key, value in vars(entity).items():
			if isinstance(value, str):
				setattr(entity, key, value.encode('utf8'))

File: sdk/javascript/src/RuleBasedTransactionFactory.js
import BaseValue from './BaseValue.js';
import ByteArray from './ByteArray.js';
import TransactionDescriptorProcessor from './TransactionDescriptorProcessor.js';

const buildEnumStringToValueMap = EnumClass => new Map(Object.getOwnPropertyNames(EnumClass)
	.filter(name => name.toUpperCase() === name)
	.map(name => [name.toLowerCase(), EnumClass[name]]));

const nameToEnumValue = (mapping, enumType, enumValueName) => {
	if (!mapping.has(enumValueName))
		throw RangeError(`unknown value ${enumValueName} for type ${enumType}`);

	return mapping.get(enumValueName);
};

const buildTypeHintsMap = structValue => {
	/** @type {{[key: string]: string}} */
	const typeHints = {};
	const rawTypeHints = structValue.constructor.TYPE_HINTS || {};
	Object.getOwnPropertyNames(rawTypeHints).forEach(key => {
		const hint = rawTypeHints[key];
		let ruleName;
		if (0 === hint.indexOf('array['))
			ruleName = hint;
		else if (0 === hint.indexOf('enum:'))
			ruleName = hint.substring('enum:'.length);
		else if (0 === hint.indexOf('pod:'))
			ruleName = hint.substring('pod:'.length);
		else if (0 === hint.indexOf('struct:'))
			ruleName = hint;

		if (ruleName)
			typeHints[key] = ruleName;
	});

	return typeHints;
};

const typeConverterFactory = (module, customTypeConverter, value) => {
	if (customTypeConverter && customTypeConverter(value))
		return customTypeConverter(value);

	if (value instanceof ByteArray)
		return new module[value.constructor.name](value.bytes);

	return value;
};

const autoEncodeStrings = entity => {
	Object.getOwnPropertyNames(entity).forEach(key => {
		const value = entity[key];
		if ('string' === typeof (value))
			entity[key] = new TextEncoder().encode(value);
	});
};

/**
 * Rule based transaction factory.
 * @note This class is not intended to be used directly.
 */
export default class RuleBasedTransactionFactory {
	/**
	 * Creates a rule based transaction factory for use with catbuffer generated code.
	 * @param {object} module Catbuffer generated module.
	 * @param {function|undefined} typeConverter Type converter.
	 * @param {Map<string, function>|undefined} typeRuleOverrides Type rule overrides.
	 */
	constructor(module, typeConverter = undefined, typeRuleOverrides = undefined) {
		/**
		 * @private
		 */
		this._module = module;

		/**
		 * Tries to coerce a value to a more appropriate type.
		 * @param {object} value Original value.
		 * @returns {object} Type converted value.
		 * @private
		 */
		this._typeConverter = value => typeConverterFactory(this._module, typeConverter, value);

		/**
		 * @private
		 */
		this._typeRuleOverrides = typeRuleOverrides || new Map();

		/**
		 * Map of rule names to transform functions.
		 * @type Map<string, function>
		 */
		this.rules = new Map();
	}

	/**
	 * Looks up a class in the wrapped module.
	 * @param {string} name Class name.
	 * @returns {Constructable} Class type.
	 * @private
	 */
	_getModuleClass(name) {
		return this._module[name];
	}

	/**
	 * Creates wrapper for SDK POD types.
	 * @param {string} name Class name.
	 * @param {Constructable} PodClass Class type.
	 */
	addPodParser(name, PodClass) {
		if (this._typeRuleOverrides.has(PodClass)) {
			this.rules.set(name, this._typeRuleOverrides.get(PodClass));
			return;
		}

		this.rules.set(name, value => (value instanceof PodClass ? value : new PodClass(value)));
	}

	/**
	 * Creates flag type parser.
	 * @param {string} name Class name.
	 */
	addFlagsParser(name) {
		const FlagsClass = this._getModuleClass(name);
		const stringToEnum = buildEnumStringToValueMap(FlagsClass);

		this.rules.set(name, flags => {
			if ('string' === typeof (flags)) {
				const enumArray = flags.split(' ').map(flagName => nameToEnumValue(stringToEnum, name, flagName));
				return new FlagsClass(enumArray.map(flag => flag.value).reduce((x, y) => x | y));
			}

			if ('number' === typeof (flags) && Number.isInteger(flags))
				return new FlagsClass(flags);

			return flags;
		});
	}

	/**
	 * Creates enum type parser.
	 * @param {string} name Class name.
	 */
	addEnumParser(name) {
		const EnumClass = this._getModuleClass(name);
		const stringToEnum = buildEnumStringToValueMap(EnumClass);

		this.rules.set(name, enumValue => {
			if ('string' === typeof (enumValue))
				return nameToEnumValue(stringToEnum, name, enumValue);

			if ('number' === typeof (enumValue) && Number.isInteger(enumValue))
				return new EnumClass(enumValue);

			return enumValue;
		});
	}

	/**
	 * Creates struct parser (to allow nested parsing).
	 * @param {string} name Class name.
	 */
	addStructParser(name) {
		const StructClass = this._getModuleClass(name);

		this.rules.set(`struct:${name}`, structDescriptor => {
			const structProcessor = this._createProcessor(structDescriptor);
			const structValue = new StructClass();

			const allTypeHints = buildTypeHintsMap(structValue);
			structProcessor.setTypeHints(allTypeHints);

			structProcessor.copyTo(structValue);
			return structValue;
		});
	}

	/**
	 * Creates array type parser, based on some existing element type parser.
	 * @param {string} name Class name.
	 */
	addArrayParser(name) {
		const elementRule = this.rules.get(name);
		if (!elementRule)
			throw Error(`cannot create array type parser because element rule "${name}" is unknown`);

		const elementName = name.replace(/^struct:/, '');
		this.rules.set(`array[${elementName}]`, values => values.map(value => elementRule(value)));
	}

	/**
	 * Autodetects rules using reflection.
	 */
	autodetect() {
		Object.getOwnPropertyNames(this._module).forEach(key => {
			const cls = this._module[key];
			if (Object.prototype.isPrototypeOf.call(BaseValue.prototype, cls.prototype))
				this.addPodParser(key, cls);
		});
	}

	/**
	 * Creates an entity from a descriptor using a factory.
	 * @param {function} factory Factory function.
	 * @param {object} descriptor Entity descriptor.
	 * @returns {object} Newly created entity.
	 */
	createFromFactory(factory, descriptor) {
		const processor = this._createProcessor(descriptor);
		const entityType = processor.lookupValue('type');
		const entity = factory(entityType);

		const allTypeHints = buildTypeHintsMap(entity);
		processor.setTypeHints(allTypeHints);
		processor.copyTo(entity, ['type']);

		autoEncodeStrings(entity);
		return entity;
	}

	/**
	 * Creates a transaction descriptor processor around a descriptor.
	 * @param {object} descriptor Transaction descriptor.
	 * @returns {TransactionDescriptorProcessor} Transaction descriptor processor.
	 * @private
	 */
	_createProcessor(descriptor) {
		return new TransactionDescriptorProcessor(descriptor, this.rules, this._typeConverter);
	}
}

// region type declarations

/**
 * Constructable class type.
 * @class
 * @typedef {{new(...args: any[]): object}} Constructable
 */

// endregion
